test/test_jit.py

```diff
@@ -1755,14 +1771,124 @@ def forward(self, x):
+                else:
+                    scripted_or_traced = torch.jit.script(eager).copy()
+                torch._C._jit_pass_dedup_module_uses(scripted_or_traced ._c)
+                folded = wrap_cpp_module(torch._C._jit_pass_fold_convbn(scripted_or_traced ._c))
+                x = torch.rand(1, 5, 6, 6)
+                self.assertEqual(eager(x), scripted_or_traced (x))
+
+    def test_foldbn_complex_cases(self):
+        # Test that we find Conv-BN patterns in submodules
+        class SubModule(torch.nn.Module):
+            def __init__(self, num_blocks, enable_bias, enable_affine):
+                super(SubModule, self).__init__()
+                layers = []
+                for i in range(num_blocks):
+                    layers.append(torch.nn.Conv2d(20, 20, 5, 1, bias=enable_bias))
+                    bn_obj = torch.nn.BatchNorm2d(num_features=20, affine=enable_affine)
+                    if enable_affine:
+                        bn_obj.weight = torch.nn.Parameter(torch.rand_like(bn_obj.weight))
+                        bn_obj.bias = torch.nn.Parameter(torch.rand_like(bn_obj.bias))
+                    bn_obj.running_mean = torch.rand_like(bn_obj.running_mean)
+                    bn_obj.running_var = torch.rand_like(bn_obj.running_var)
+                    layers.append(bn_obj)
+                self.layers = nn.Sequential(*layers)
+
+            def forward(self, x):
+                return self.layers(x)
+
+        class TestModule(torch.nn.Module):
+            def __init__(self, num_blocks, enable_bias, enable_affine):
+                super(TestModule, self).__init__()
+                self.sub = SubModule(num_blocks, enable_bias, enable_affine)
+
+            def forward(self, x):
+                x = self.sub(x)
+                return x
+
+        bias_affine_options = itertools.product([True, False], [True, False], [True, False], [1, 2])
+        for (tracing_mode, enable_bias, enable_bn_affine, num_layers) in bias_affine_options:
+            eager = TestModule(num_layers, enable_bias, enable_bn_affine)
+            eager.eval()
+
+            if tracing_mode:
+                x = torch.rand(1, 20, 10, 10)
+                scripted_or_traced = torch.jit.trace(eager, x)
+            else:
+                scripted_or_traced = torch.jit.script(eager)
+            scripted_or_traced.eval()
+
+            FileCheck().check_count("prim::CallMethod[name=\"forward\"]", num_layers * 2, exactly=True) \
+                .run(str(get_forward_graph(scripted_or_traced.sub.layers._c)))
+
+            scripted_or_traced = wrap_cpp_module(torch._C._jit_pass_fold_convbn(scripted_or_traced._c))
+
+            FileCheck().check_count("prim::CallMethod[name=\"forward\"]", num_layers, exactly=True) \
+                .run(str(get_forward_graph(scripted_or_traced.sub.layers._c)))
+
+            x = torch.rand(1, 20, 10, 10)
+            self.assertEqual(eager(x), scripted_or_traced(x))
+
+    def test_nofoldbn_complex_cases(self):
```

Comment:
what is the difference between this test and the previous test?

Reply:
This adds `ReLU` in between so that folding cannot take place. I am just testing that we dont fold. If you dont think this is needed, I can remove it. It may add to test time actually, otherwise.

Reply:
yeah I think we don't need this. this sounds like some general case that won't be matched and there will be variety of these cases. Unless we have some specific corner case that we want to test, e.g. user defiend class ends with Conv2d, probably not worth to write negative tests